lib/types/message.js

```diff
@@ -1,347 +1,191 @@
-  Properties.super_.call(this, Properties);
-
-  u.assignDefined(this, options, {
-    messageId: options.messageId,
-    userId: u.coerce(options.userId, Buffer),
-    to: options.to,
-    subject: options.subject,
-    replyTo: options.replyTo,
-    correlationId: options.correlationId,
-    contentType: options.contentType,
-    contentEncoding: options.contentEncoding,
-    absoluteExpiryTime: options.absoluteExpiryTime,
-    creationTime: options.creationTime,
-    groupId: options.groupId,
-    groupSequence: options.groupSequence,
-    replyToGroupId: options.replyToGroupId
-  });
+  defineSection(descriptor, sectionDefinition);
+  sectionByProperty[property] = sectionDefinition;
 }
 
-util.inherits(Properties, DescribedType);
-
-Properties.prototype.Descriptor = { code: 0x73, name: 'amqp:properties:list' };
-Properties.prototype.EncodeOrdering = [
-  'messageId', 'userId', 'to', 'subject', 'replyTo', 'correlationId',
-  'contentType', 'contentEncoding', 'absoluteExpiryTime', 'creationTime',
-  'groupId', 'groupSequence', 'replyToGroupId'
-];
-
-Properties.fromDescribedType = function(describedType) {
-  var options = {};
-  u.assignFromDescribedType(Properties, describedType, options);
-
-  ['contentType', 'contentEncoding'].forEach(function(option) {
-    if (options[option] instanceof AMQPSymbol)
-      options[option] = options[option].contents;
-  });
-
-  return new Properties(options);
-};
-
-Properties.prototype.getValue = function() {
-  var self = this;
-  return {
-    messageId: u.orNull(self.messageId),
-    userId: u.orNull(self.userId),
-    to: u.orNull(self.to),
-    subject: u.orNull(self.subject),
-    replyTo: u.orNull(self.replyTo),
-    correlationId: u.orNull(self.correlationId),
-    contentType: (self.contentType === undefined) ? null : new AMQPSymbol(self.contentType),
-    contentEncoding: (self.contentEncoding === undefined) ? null : new AMQPSymbol(self.contentEncoding),
-    absoluteExpiryTime: (self.absoluteExpiryTime === undefined) ? null : new ForcedType('timestamp', self.absoluteExpiryTime),
-    creationTime: (self.creationTime === undefined) ? null : new ForcedType('timestamp', self.creationTime),
-    groupId: u.orNull(self.groupId),
-    groupSequence: (self.groupSequence === undefined) ? null : new ForcedType('uint', self.groupSequence),
-    replyToGroupId: u.orNull(self.replyToGroupId),
-    encodeOrdering: Properties.prototype.EncodeOrdering
+function defineCompositeSection(definition) {
+  var descriptor = { name: 'amqp:' + definition.name + ':list', code: definition.code };
+  var sectionDefinition = {
+    descriptor: definition.code,
+    encode: function(value, buffer) {
+      var _len = definition.fields.length, result = [];
+      for (var i = 0; i < _len; ++i) {
+        var field = definition.fields[i];
+        if (value && value.hasOwnProperty(field.name)) {
+          result[i] = field.type !== '*' ? new ForcedType(field.type, value[field.name]) : value[field.name];
```

Comment:
should these be a `wrapField` call in case value[field.name] is already the correct type?